From 92323cf4b506c40376be74e955836da30980ae54 Mon Sep 17 00:00:00 2001
From: Donald Sharp <sharpd@nvidia.com>
Date: Wed, 13 Mar 2024 10:26:58 -0400
Subject: [PATCH 1/3] bgpd: Ensure that the correct aspath is free'd

Currently in subgroup_default_originate the attr.aspath
is set in bgp_attr_default_set, which hashs the aspath
and creates a refcount for it.  If this is a withdraw
the subgroup_announce_check and bgp_adj_out_set_subgroup
is called which will intern the attribute.  This will
cause the the attr.aspath to be set to a new value
finally at the bottom of the function it intentionally
uninterns the aspath which is not the one that was
created for this function.  This reduces the other
aspath's refcount by 1 and if a clear bgp * is issued
fast enough the aspath for that will be removed
and the system will crash.

Signed-off-by: Donald Sharp <sharpd@nvidia.com>
---
diff --git a/bgpd/bgp_route.c b/bgpd/bgp_route.c
index fc776a4fd..a0113c04e 100644
--- a/bgpd/bgp_route.c
+++ b/bgpd/bgp_route.c
@@ -3012,7 +3012,7 @@ void subgroup_process_announce_selected(struct update_subgroup *subgrp,
 				   PEER_STATUS_ORF_WAIT_REFRESH))
 		return;

-	memset(&attr, 0, sizeof(attr));
+	memset(pattr, 0, sizeof(*pattr));
 	/* It's initialized in bgp_announce_check() */

 	/* Announcement to the subgroup. If the route is filtered withdraw it.
